curp/src/sync_manager.rs

```diff
@@ -144,137 +137,222 @@ impl<C: Command + 'static> SyncManager<C> {
+        // send append_entries request and receive response
+        debug!("append_entries sent to {}", connect.addr);
+        let resp = connect
+            .append_entries(req, Self::APPEND_ENTRIES_TIMEOUT)
+            .await;
+
+        #[allow(clippy::unwrap_used)]
+        // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+        match resp {
+            Err(e) => warn!("Heartbeat error: {}", e),
+            Ok(resp) => {
+                // TODO: update term
+                let resp = resp.into_inner();
+                let mut state = state.write();
+                if resp.success {
+                    // update match_index and next_index
+                    *state.match_index.get_mut(&connect.addr).unwrap() = last_sent_index;
+                    *state.next_index.get_mut(&connect.addr).unwrap() = last_sent_index + 1;
+
+                    // update commit_index if needed
+                    let majority_cnt = (state.others.len() + 1) / 2 + 1;
+                    for i in ((state.commit_index + 1)..=state.last_log_index()).rev() {
+                        // should not commit logs from previous term
+                        if state.log[i].term() != state.term {
+                            break;
+                        }
+
+                        // calculate the number of followers that has matched log[i]
+                        let match_count = state
+                            .others
+                            .iter()
+                            .filter(|addr| state.match_index[*addr] >= i)
+                            .count();
+
+                        // If the majority of servers has replicated the log, commit
+                        // + 1 means including the leader itself
+                        if match_count + 1 >= majority_cnt {
+                            debug!("commit_index updated to {i}");
+                            state.commit_index = i;
+                            commit_notify.notify_one();
+                            ae_notify.notify_one();
+                            break;
+                        }
+                    }
+                } else {
+                    *state.next_index.get_mut(&connect.addr).unwrap() -= 1;
+                }
+            }
+        };
+    }
+
+    /// Background apply
+    async fn background_apply<CE: 'static + CommandExecutor<C>>(
+        state: Arc<RwLock<State<C>>>,
+        cmd_executor: Arc<CE>,
+        comp_chan: SpmcKeyBasedSender<C::K, SyncCompleteMessage<C>>,
+        commit_notify: Arc<Notify>,
+    ) {
+        loop {
+            commit_notify.notified().await;
+            if state.read().need_commit() {
+                let mut state = state.write();
+                #[allow(clippy::integer_arithmetic, clippy::indexing_slicing)]
+                // TODO: will last_applied overflow?
+                for i in (state.last_applied + 1)..=state.commit_index {
```

Comment:
Yes, we should do GC to remove the useless entries or store them on the disk. So the last_applied won't overflow.